node/pkg/libp2p/setup/setup.go

```diff
@@ -0,0 +1,209 @@
+		}
+	}
+
+	h, err := makeHost(listenPort, priv)
+	if err != nil {
+		log.Error().Err(err).Msg("Error creating libp2p host")
+		return nil, err
+	}
+
+	_, err = utils.InitDHT(ctx, h, "")
+	if err != nil {
+		log.Error().Err(err).Msg("Error initializing DHT")
+		return nil, err
+	}
+
+	pi := peer.AddrInfo{
+		ID:    h.ID(),
+		Addrs: h.Addrs(),
+	}
+
+	for _, addr := range pi.Addrs {
+		fmt.Println(addr.String() + "/p2p/" + h.ID().String())
+	}
+
+	return &h, nil
+}
+
+func Setup(ctx context.Context, bootnodeStr string, port int) (*host.Host, *pubsub.PubSub, error) {
+	host, err := MakeHost(port)
+	if err != nil {
+		return nil, nil, err
+	}
+	ps, err := MakePubsub(ctx, host)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if bootnodeStr != "" {
+		bootnode, bootErr := multiaddr.NewMultiaddr(bootnodeStr)
+		if bootErr != nil {
+			return nil, nil, bootErr
+		}
+
+		peerinfo, bootErr := peer.AddrInfoFromP2pAddr(bootnode)
+		if bootErr != nil {
+			return nil, nil, bootErr
+		}
+
+		bootErr = host.Connect(ctx, *peerinfo)
+		if bootErr != nil {
+			return nil, nil, bootErr
+		}
+	}
+
+	discoverString := "orakl-topic-discovery-2024"
+	go func() {
+		if err = utils.DiscoverPeers(ctx, host, discoverString, bootnodeStr); err != nil {
+			log.Error().Err(err).Msg("Error from DiscoverPeers")
+		}
+	}()
+
+	return &host, ps, nil
+}
+
+func SetupFromBootApi(ctx context.Context, port int) (host.Host, *pubsub.PubSub, error) {
+	host, err := MakeHost(port)
+	if err != nil {
+		log.Error().Err(err).Msg("Error making host")
+		return nil, nil, err
+	}
+
+	ps, err := MakePubsub(ctx, host)
+	if err != nil {
+		log.Error().Err(err).Msg("Error making pubsub")
+		return nil, nil, err
+	}
+
+	ip, port, hostId, err := utils.ExtractPayloadFromHost(host)
+	if err != nil {
+		log.Error().Err(err).Msg("Error extracting payload from host")
+		return nil, nil, err
+	}
+
+	apiEndpoint := os.Getenv("BOOT_API_URL")
+	if apiEndpoint == "" {
+		log.Info().Msg("boot api endpoint not set, using default url: http://localhost:8089")
+		apiEndpoint = "http://localhost:8089"
+	}
+
+	dbPeers, err := request.UrlRequest[[]BootPeerModel](apiEndpoint+"/api/v1/peer/sync", "POST", map[string]any{
+		"ip":      ip,
+		"port":    port,
+		"host_id": hostId,
+	}, nil, "")
+	if err != nil {
+		log.Error().Err(err).Msg("Error getting peers from boot API")
+		return nil, nil, err
+	}
+
+	for _, dbPeer := range dbPeers {
+		peerAddr := fmt.Sprintf("/ip4/%s/tcp/%d/p2p/%s", dbPeer.Ip, dbPeer.Port, dbPeer.HostId)
+		peerMultiAddr, err := multiaddr.NewMultiaddr(peerAddr)
+		if err != nil {
+			log.Error().Err(err).Msg("Error creating multiaddr")
+			return nil, nil, err
+		}
+
+		peerInfo, err := peer.AddrInfoFromP2pAddr(peerMultiAddr)
+		if err != nil {
+			log.Error().Err(err).Msg("Error getting AddrInfo from p2p address")
+			return nil, nil, err
+		}
+
+		err = host.Connect(ctx, *peerInfo)
+		if err != nil {
+			log.Error().Err(err).Msg("error connecting to peer")
+			return nil, nil, err
+		}
+	}
```

Comment:
considering if I just should continue if err occurs rather than returning immediately